Add OpenVino Detector

NateMeyer opened the pull request:
This PR is dependent on the Object Detector Abstraction change #3656.

This adds the OpenVino framework and a default model to the build image and a detector implementation that uses the OpenVino framework.  I have developed and tested this on my i5 1135G7, using the Xe iGPU.

The necessary configuration to run the included model is explained in the Detectors page of the documentation.

TODO:

- [x] Build OpenVino Wheel for amd64
- [x] Build OpenVino Wheel for ARM (32 and 64)
- [x] Determine what is needed to run on VPU/MyriadX hardware
- [x] Rebase to dev branch
- [x] Update Documentation with supported hardware

~- Update default model~ Further Model support can be reviewed/developed later

NateMeyer committed:
Initial work for adding OpenVino detector.  Not functional

NateMeyer committed:
Load model and submit for inference.

Sucessfully load model and initialize OpenVino engine with either CPU or GPU as device.
Does not parse results for objects.

NateMeyer committed:
Detection working with ssdlite_mobilenetv2 FP16 model

NateMeyer committed:
Add OpenVIno support and model to docker image

NateMeyer committed:
Add documentation for OpenVino detector configuration

NickM-27 commented on docker/Dockerfile:
```diff
@@ -42,6 +42,20 @@ RUN pip3 install -r requirements.txt
 
+# Download and Convert OpenVino model
+FROM debian:11 as ov-converter
+RUN apt-get update && apt-get install -y wget ffmpeg libsm6 libxext6 \
```
why is ffmpeg needed here?

NateMeyer replied:
Good question.  I think that was carried over from one of my dev images.  I'll review the packages included in the converter stage and the final image and try to pare it back.

NateMeyer replied:
I updated the packages for this converter stage and merged the intel-opencl-icd package into the arm64-only section of the final stage.

NateMeyer commented:
I have two questions with this:

1. I've only tested the CPU and GPU devices with this.  Is there someone who has a NCS2 that can try the VPU device?
2. When looking through the docker build packages, I realize I don't know if or how this will build for arm/arm64 architectures.  I believe the CPU and GPU devices only work for Intel parts (i.e. amd64 arch), but in theory the VPU should work on an arm board.  I tried briefly to run a multiarch build, but would I need to do some more setup to get that to build.

NickM-27 commented:
> 1. I've only tested the CPU and GPU devices with this.  Is there someone who has a NCS2 that can try the VPU device?

I know a couple users have mentioned they have one, not sure which ones will be able to test, you might want to make a discussion post about it.

> 2. When looking through the docker build packages, I realize I don't know if or how this will build for arm/arm64 architectures.  I believe the CPU and GPU devices only work for Intel parts (i.e. amd64 arch), but in theory the VPU should work on an arm board.  I tried briefly to run a multiarch build, but would I need to do some more setup to get that to build.

It's not clear to me what you're asking. Are you saying the multiarch build fails somewhere? Are all the dependencies needed for the VPU in the ov package that is added? 

If not then I'd imagine you'd just need to add the `intel-opencl-icd` to arm as well

NickM-27 commented on docker/Dockerfile:
```diff
@@ -101,6 +104,7 @@ RUN apt-get -qq update \
     # arch specific packages
     && if [ "${TARGETARCH}" = "amd64" ]; then \
     apt-get -qq install --no-install-recommends --no-install-suggests -y \
+    intel-opencl-icd \
```
If this is specifically for openvino then I would recommend putting a comment that explains that so it is clear 👍

NateMeyer commented:
> > 2. When looking through the docker build packages, I realize I don't know if or how this will build for arm/arm64 architectures.  I believe the CPU and GPU devices only work for Intel parts (i.e. amd64 arch), but in theory the VPU should work on an arm board.  I tried briefly to run a multiarch build, but would I need to do some more setup to get that to build.
> 
> It's not clear to me what you're asking. Are you saying the multiarch build fails somewhere? Are all the dependencies needed for the VPU in the ov package that is added?
> 
> If not then I'd imagine you'd just need to add the `intel-opencl-icd` to arm as well

The intel-opencl-icd package is only needed for running on the Intel GPU, and is only available for the amd64 arch.

The multiarch build is failing for me, I assume it is just a setup issue on my end.  Are there some instructions for running the multiarch build? My quick googling looks like I need to setup qemu on my system to run the ARM builds.

I think the openvino python library should have the necessary pieces for using the VPU plugin, except for the driver which would have to be installed on the host system.  I would then expect the enumerated VPU device could be passed through to the container.

NickM-27 commented:
You will want to use https://github.com/blakeblackshear/frigate/blob/release-0.11.0/Makefile#L29 to build multiarch. If there is an error then you'll need to paste it here and I can advise.

NateMeyer commented:
The error I get is:
```
 => ERROR [wheels 2/8] RUN apt-get -qq update     && apt-get -qq install -y     apt-transport-http  0.4s
 => CANCELED [ov-converter 2/8] RUN apt-get -qq update     && apt-get -qq install -y wget python3   0.5s
------
 > [wheels 2/8] RUN apt-get -qq update     && apt-get -qq install -y     apt-transport-https     gnupg     wget     && apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 9165938D90FDDD2E     && echo "deb http://raspbian.raspberrypi.org/raspbian/ bullseye main contrib non-free rpi" | tee /etc/apt/sources.list.d/raspi.list     && apt-get -qq update     && apt-get -qq install -y     python3     python3-dev     wget     build-essential cmake git pkg-config libgtk-3-dev     libavcodec-dev libavformat-dev libswscale-dev libv4l-dev     libxvidcore-dev libx264-dev libjpeg-dev libpng-dev libtiff-dev     gfortran openexr libatlas-base-dev libssl-dev    libtbb2 libtbb-dev libdc1394-22-dev libopenexr-dev     libgstreamer-plugins-base1.0-dev libgstreamer1.0-dev     gcc gfortran libopenblas-dev liblapack-dev:
#0 0.355 exec /bin/sh: exec format error
------
error: failed to solve: executor failed running [/bin/sh -c apt-get -qq update     && apt-get -qq install -y     apt-transport-https     gnupg     wget     && apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 9165938D90FDDD2E     && echo "deb http://raspbian.raspberrypi.org/raspbian/ bullseye main contrib non-free rpi" | tee /etc/apt/sources.list.d/raspi.list     && apt-get -qq update     && apt-get -qq install -y     python3     python3-dev     wget     build-essential cmake git pkg-config libgtk-3-dev     libavcodec-dev libavformat-dev libswscale-dev libv4l-dev     libxvidcore-dev libx264-dev libjpeg-dev libpng-dev libtiff-dev     gfortran openexr libatlas-base-dev libssl-dev    libtbb2 libtbb-dev libdc1394-22-dev libopenexr-dev     libgstreamer-plugins-base1.0-dev libgstreamer1.0-dev     gcc gfortran libopenblas-dev liblapack-dev]: exit code: 1
make: *** [Makefile:25: arm64] Error 1

```

when it gets to the first `RUN` step in the arm64 target.

It is complaining that it can't run the arm binaries in the arm image.  I just installed qemu and followed the setup at https://github.com/multiarch/qemu-user-static and got past this first error.   To prove out my build system, I built the release-0.11.0 branch.  This successfully built each architecture.  Finally I had to create a buildx builder to create the multiarch image at the end with `docker buildx create --name localBuilder --use`.  After all that, the `build` make target finishes.

Back on my openvino branch, now I get to the next issue. The Python package available from pypi only support amd64 architecture. 

In order to support OpenVino on arm, it looks like we'll have to add another stage to the build to compile the runtime from source, until someone publishes a python package.  The repository from Intel will build on ARM, but only supports the VPU device. [OpenVino Build Instructions for RaspbianOS](https://github.com/openvinotoolkit/openvino/wiki/BuildingForRaspbianStretchOS)

There is another plugin that supports the ARM CPU for OpenVIno that is maintained by OpenCV, but I'm not sure that will bring much value to this project. [OpenVino ARM CPU Plugin Repo](https://github.com/openvinotoolkit/openvino_contrib/tree/master/modules/arm_plugin)

At this point, I think I'll keep this to amd64 only.  If there is strong desire to add the build of the openvino runtime on arm, we can look at adding it then.

NateMeyer committed:
Adds support for ARM32/ARM64 and the Myriad X hardware

-  Use custom-built openvino wheel for all platforms
-  Add libusb build without udev for NCS2 support

NickM-27 commented:
Yeah, the NCS2 is already end-of-lifed so I don't know if that makes sense to worry about.

NickM-27 commented on docker/Dockerfile:
```diff
@@ -40,24 +40,36 @@ COPY requirements.txt /requirements.txt
 RUN pip3 install -r requirements.txt
 
 COPY requirements-wheels.txt /requirements-wheels.txt
+COPY requirements-wheels-amd64.txt /requirements-wheels-amd64.txt
```
We don't need to copy this unless the build is amd64

NateMeyer replied:
I am not aware of a way to condition this `COPY` by platform in the Dockerfile.  This is only an intermediate stage and the file shouldn't be in the final image.

NickM-27 replied:
Yeah I was being silly, disregard that.

NickM-27 commented on docker/Dockerfile:
```diff
@@ -40,24 +40,36 @@ COPY requirements.txt /requirements.txt
-    && python3 get-pip.py "pip"
-
+# Install OpenVino Runtime and Dev library
 COPY requirements-ov.txt /requirements-ov.txt
```
Don't need to copy this unless amd64

NickM-27 commented:
By the way what inference times are you seeing?

NateMeyer commented:
The model benchmark in the OpenVino toolbox claims my GPU will do 263fps with this model.  Looks like the debug page is showing an inference speed of 8-12ms.  

This is just measuring latency? My GPU is barely loaded during this, but I am only running with a single camera at the moment.  The measure of throughput could be higher than the inference speed suggests if the async requests or batching are played with.

NickM-27 commented:
The debug page is the average amount of time it has taken to start the detection then get a result. Maybe with more cameras it could be loaded higher but that is quite good.

yeahme49 commented:
Just wanted to say I tested this on my i5-4590 using the CPU plugin and I'm getting 15-18ms with 10-30% CPU usage (2 cores on a Proxmox LXC container). Much better than the built-in CPU plugin, I would get around 80-100ms (sometimes as high as 200ms) and 80% CPU usage. I believe my iGPU is unsupported so I can't test that (plus I think it's disabled due to having a dedicated GPU installed). 

This plugin gets about the same, if not slightly better, speed as the TensorRT CUDA plugin with my P400 card (https://github.com/blakeblackshear/frigate/pull/2548 / https://github.com/blakeblackshear/frigate/discussions/3016).

I definitely hope this gets merged into Frigate, awesome job on this!

NickM-27 commented on docs/docs/configuration/detectors.md:
```diff
@@ -77,3 +77,33 @@ detectors:
+
+### OpenVino Devices
+
+The OpenVino detector supports the Intel-supplied device plugins and can specify one or more devices in the configuration.  See OpenVino's device naming conventions in the [Device Documentation](https://docs.openvino.ai/latest/openvino_docs_OV_UG_Working_with_devices.html) for more detail. Other supported devices could be `AUTO`, `CPU`, `GPU`, etc.
```
I assume you may be waiting for results from your test, but just confirming you intend to add a section for the `VPU` NCS2 if it works out?

NateMeyer replied:
Yes, I didn't want to add documentation I had no way of verifying was correct.

NickM-27 commented:
> (plus I think it's disabled due to having a dedicated GPU installed).

Assuming you're running on linux it doesn't work that, they'll both work just fine together. What error were you getting when trying it with the GPU?

@NateMeyer We will definitely need documentation on which intel iGPUs are supported for this.

NateMeyer commented:
Yea, it should support GPUs per Intel's "Supported Devices" documentation but we can be explicit about it too.

This still needs to either add ARM support for openvino, or add more platform checks in the detector code. As it is, it will crash on the arm64 and armv7 builds.

yeahme49 commented:
> > (plus I think it's disabled due to having a dedicated GPU installed).
> 
> Assuming you're running on linux it doesn't work that, they'll both work just fine together. What error were you getting when trying it with the GPU?
> 
> @NateMeyer We will definitely need documentation on which intel iGPUs are supported for this.
```
frigatevino    | Process detector:vino:
frigatevino    | Traceback (most recent call last):
frigatevino    |   File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
frigatevino    |     self.run()
frigatevino    |   File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
frigatevino    |     self._target(*self._args, **self._kwargs)
frigatevino    |   File "/opt/frigate/frigate/object_detection.py", line 121, in run_detector
frigatevino    |     object_detector = LocalObjectDetector(
frigatevino    |   File "/opt/frigate/frigate/object_detection.py", line 66, in __init__
frigatevino    |     self.detect_api = OvDetector(
frigatevino    |   File "/opt/frigate/frigate/detectors/openvino.py", line 16, in __init__
frigatevino    |     self.interpreter = self.ov_core.compile_model(
frigatevino    |   File "/usr/local/lib/python3.9/dist-packages/openvino/runtime/ie_api.py", line 266, in compile_model
frigatevino    |     super().compile_model(model, device_name, {} if config is None else config)
frigatevino    | RuntimeError: Failed to create plugin /usr/local/lib/python3.9/dist-packages/openvino/libs/libopenvino_intel_gpu_plugin.so for device GPU
frigatevino    | Please, check your environment
frigatevino    | [CLDNN ERROR]. clGetPlatformIDs error -1001
```

If I use the OpenVIno hello_device_query.py sample it only shows CPU, no GPU device
```
[ INFO ] Available devices:
[ INFO ] CPU :
[ INFO ]        SUPPORTED_PROPERTIES:
[ INFO ]                AVAILABLE_DEVICES:
[ INFO ]                RANGE_FOR_ASYNC_INFER_REQUESTS: 1, 1, 1
[ INFO ]                RANGE_FOR_STREAMS: 1, 2
[ INFO ]                FULL_DEVICE_NAME: Intel(R) Core(TM) i5-4590 CPU @ 3.30GHz
[ INFO ]                OPTIMIZATION_CAPABILITIES: FP32, FP16, INT8, BIN, EXPORT_IMPORT
[ INFO ]                CACHE_DIR:
[ INFO ]                NUM_STREAMS: 1
[ INFO ]                AFFINITY: UNSUPPORTED TYPE
[ INFO ]                INFERENCE_NUM_THREADS: 0
[ INFO ]                PERF_COUNT: False
[ INFO ]                INFERENCE_PRECISION_HINT: UNSUPPORTED TYPE
[ INFO ]                PERFORMANCE_HINT: UNSUPPORTED TYPE
[ INFO ]                PERFORMANCE_HINT_NUM_REQUESTS: 0
[ INFO ]
```

I also tried running one of OpenVino's docker images that have all the runtimes and drivers installed and I get the same result. It looks like Intel NEO Compute compatibility is required and that is Broadwell and above (https://github.com/intel/compute-runtime#supported-platforms)

NickM-27 commented:
@yeahme49 Perfect thanks, that will be helpful for our docs.

NickM-27 commented:
@NateMeyer Looks like we'll want to update the docs that intel GPUs are supported 5th gen and up.

NateMeyer commented:
Intel's documentation looks like 6th gen and up.  I'll get this info added after we settle out the NCS2 and ARM support.

NickM-27 commented:
> Intel's documentation looks like 6th gen and up. I'll get this info added after we settle out the NCS2 and ARM support.

Sounds good, seems like you're making headway with it.

titilambert commented:
Hello, I have a nsc2. I guess I can make a test.
Could you provide a configuration example ?
Thanks

NateMeyer commented:
> Hello, I have a nsc2. I guess I can make a test. Could you provide a configuration example ? Thanks

Hi @titilambert, there are some instructions in the discussion topic [3797](https://github.com/blakeblackshear/frigate/discussions/3797).  Just add the detector and model blocks to your normal config, and point to my test image.  Post a comment over there and I can help with any other questions. Thanks!

CarlosML commented:
@NateMeyer, thank you very much for your work! I have been waiting to be able to use OpenVINO with Frigate, since in my country it is impossible to get a Google Coral USB. Also, previously using the OpenVINO object detection demo I noticed that nanodet models were not giving me the false positives that I am having with ssdlite. As I wanted to test nanodet with Frigate but I don't know anything about how OpenVINO or Frigate works internally I modified the code to add the Model API that is in the Open Model Zoo of OpenVINO, as it abstracts the adaptation of the tensor to the model and process its output. No doubt it is inefficient, at least because although the tensor_transform function explicitly says that only works permutations B, H, W, C the nanodet models need B, C, H, W and yet when passing the tensor to the Model API model adapter the inference is performed correctly anyway. Even with the very anemic GPU of a Celeron J4025 I have a 65-70 ms latency with nanodet-plus-m-1.5x-416 (with nanodet-m-1.5x-416 I had a slightly lower latency, 50-55 ms), using OpenVINO 2022.2.0. 

In case anyone is interested in trying nanodet in this [gist](https://gist.github.com/CarlosML/581538f2c9446de9abbe1c8e633f1a21) there are the diffs of the modifications I made, in config.yml you pass model_type as "NanoDet", "NanoDetPlus" or "SSD" and you have to mount openvino-model externally in docker-compose.yml to pass the corresponding IRs (NanoDet uses a different tag file, coco_80cl.txt). I tried to modify Dockerfile to perform the conversion and add it to the Docker image, but when installing the OpenVINO Python dependencies to perform the Pytorch model conversion it starts to crash because of missing libraries (libGL). 

I think it is a great advantage of Frigate to support OpenVINO because besides supporting several generations of iGPUs the latest version announces Intel Arc support.

NateMeyer commented:
Hi @CarlosML, I'm glad you've found it useful!  I played with a few different models when I was working on this but I don't think I tried NanoDet.  My intent with this detector would be that you convert a model to the IR format, and you can configure the input format in the config (BCHW vs BHWC, etc).  I haven't yet added configurable processing on the detection output.  You've worked around this with using the Model API from OMZ.  This looks like a neat API, I hadn't considered using it before.

Ideally we could have the shape of the results be configurable and generic enough that the description in the model config would apply to different detector frameworks (i.e. Tensorflow-lite, RKNN, etc.).  I'm open to suggestions on how that would look.  SSD, NanoDet, and Yolo have fairly different output tensors that need to be parsed a bit differently.  Perhaps someone could add a model-parse utility with common functions to get labeled detection boxes from the various different model types.  I think these could be done in a detector-agnostic way.

CarlosML commented:
One issue that I don't know how important is that Frigate by default uses SSDLite Mobiledet which I understand has a higher accuracy score than MobileNetV2+SSDLite which is in OMZ. I understand [this one](http://download.tensorflow.org/models/object_detection/ssdlite_mobiledet_cpu_320x320_coco_2020_05_19.tar.gz) would be similar to the one Frigate downloads from the Coral repository (minus some post processing), late last year I tried to convert it with OpenVINO and it failed and still fails 'automatically' with the latest version but the [suggestion](https://github.com/openvinotoolkit/openvino/issues/9461#issuecomment-1002055436) I was given in the OpenVINO repository seems to work fine. [Here](https://github.com/CarlosML/ssdlite_mobiledet/tree/main) is the output of mo.py if anybody is interested in trying it.

BradleyFord commented:
> Yeah, the NCS2 is already end-of-lifed so I don't know if that makes sense to worry about.

True, but this croud funding campaign just launched which is based on the same processer.
https://www.crowdsupply.com/eyecloud/openncc-ncb

NateMeyer commented:
Intel says they are deprecating the NCS2 support in OpenVino after 2022.3
> Depreciation Notice: 
> 
> Intel® Corporation has discontinued the Intel® Movidius™ Neural Compute Stick 2 (Intel® NCS2). Version 2022.3 LTS will be the last version of the toolkit to support Intel NCS2. 

https://www.intel.com/content/www/us/en/developer/tools/openvino-toolkit/system-requirements.html

NickM-27 commented:
Interesting, I suppose that would include this crowdfunded device as well if newer versions of the toolkit also won't support it

NateMeyer commented:
Perhaps they would move the support to the openvino_contrib repo that also has arm and Nvidia plugins?  Then at least it could be updated by the community?  That seems best case...

BradleyFord commented:
It does mention a lot about OpenVINO, I have asked them to comment on what
will happen if Intel discontinue the API

On Fri, Sep 30, 2022, 02:11 Nate Meyer ***@***.***> wrote:

> Perhaps they would move the support to the openvino_contrib repo that also
> has arm and Nvidia plugins? Then at least it could be updated by the
> community? That seems best case...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/blakeblackshear/frigate/pull/3768#issuecomment-1262705616>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACOZ72VUVTEB2B2D7FV6JDWAXSXTANCNFSM6AAAAAAQFDFYYU>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

NateMeyer commented:
I would expect Intel would still support the silicon if they're still producing/selling it.  Perhaps it falls under the more generic Myriad X support instead of being NCS2 support.

Johannasss commented:
> It does mention a lot about OpenVINO, I have asked them to comment on what will happen if Intel discontinue the API
> […](#)
> On Fri, Sep 30, 2022, 02:11 Nate Meyer ***@***.***> wrote: Perhaps they would move the support to the openvino_contrib repo that also has arm and Nvidia plugins? Then at least it could be updated by the community? That seems best case... — Reply to this email directly, view it on GitHub <[#3768 (comment)](https://github.com/blakeblackshear/frigate/pull/3768#issuecomment-1262705616)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AACOZ72VUVTEB2B2D7FV6JDWAXSXTANCNFSM6AAAAAAQFDFYYU> . You are receiving this because you commented.Message ID: ***@***.***>

Thanks @BradleyFord for the invitation. The [crowdfunding device](https://www.crowdsupply.com/eyecloud/openncc-ncb) you mentioned is our OpenNCC NCB, which is built from the Intel Movidius X VPU. Our team has been developing projects using this VPU for many years, and this device is built for those who are looking for NCS2 replacements because intel no longer builds and sells NCS2.

This NCB device can work with OpenVINO(versions 2020.3 & 2021.4 verified). We haven't tested other OpenVINO versions, and by now we have no plan to do that unless there are commercial customization requirements or volume orders. 

Feel free to correct me, from what I learned from this discussion loop, @NateMeyer mainly wants to verify working OpenVINO on a combination of [arm+ncs2], right? For your reference, we and our customer have tested and verified working OpenVINO(only v2020.3 or v2021.4) on a combination of [RaspberryPi+OpenNCC NCB], similar as using [RaspberryPi + NCS2] except replacing the original .mvcmd file under openvino/inference_engine/lib/intel64/usb-ma2x8x.mvcmd with one for OpenNCC NCB (you can find it [here](https://github.com/EyecloudAi/openncc/blob/R22.08.01/SDK/Source/Firmware/fw/usb-ma2x8x.mvcmd) ).

Feel free to let me know if you have any further questions on NCB.

NateMeyer commented:
Hi @Johannasss, thanks for the information.  For this project, I am currently targeting the 2022.2 release of OpenVINO with the newer API.  It currently works on the Myriad X and the NCS2, so it would be interesting if it ran on your NCB hardware.  Our discussions here are looking at what hardware could be supported and were wondering what would happen to the Myriad X support from Intel's future OpenVINO releases after the NCS2 is deprecated at the end of the year.

I'm a little concerned about the need for a different `mvcmd` file, as this makes it more difficult to support both the NCS2 and your hardware.  The user will likely have to map in the replacement file in their docker container.

NateMeyer commented:
It took about a month, but I have a multi-arch image that I believe will support NCS2 on x86_64 and ARM 32/64.  I just want to cleanup the Dockerfile to build libusb in a separate stage then I'll merge those changes into this PR.  I updated the NCS2 discussion with a new image if anyone is testing or wants to try out the arm builds (#3797).

NateMeyer committed:
Add documentation around Intel CPU requirements and NCS2 setup

NickM-27 commented:
Feel free to do it if you want, but just leaving my thoughts here that we will likely want to revisit the `Recommended Hardware` page in the docs to reflect the different detector options as well as the expected inference times, what that means, necessary platforms, etc.

NateMeyer commented:
> Feel free to do it if you want, but just leaving my thoughts here that we will likely want to revisit the `Recommended Hardware` page in the docs to reflect the different detector options as well as the expected inference times, what that means, necessary platforms, etc.

We'll want to get more runtime in a broader setting to get a better feel for performance before making recommendations.

Also, I'm curious on thoughts about the included model.  The one I have now is fairly fast, but I think it's not the most accurate.  I think we'll have to look at a handful of models and see what works best for people?

blakeblackshear commented:
The vast majority of models available to try will all be trained on the same COCO dataset. The primary difference will be in the size of the images it is trained on. I have tried nearly all the various model architectures mobilenet/mobiledet/spagettinet/efficientdet and I don't see much of a difference in precision unless you go with larger image size versions from efficientdet. This has a significant tradeoff in terms of inference speeds.

yeahme49 commented:
> > Feel free to do it if you want, but just leaving my thoughts here that we will likely want to revisit the `Recommended Hardware` page in the docs to reflect the different detector options as well as the expected inference times, what that means, necessary platforms, etc.
> 
> We'll want to get more runtime in a broader setting to get a better feel for performance before making recommendations.
> 
> Also, I'm curious on thoughts about the included model. The one I have now is fairly fast, but I think it's not the most accurate. I think we'll have to look at a handful of models and see what works best for people?

Is there a place to get other models for testing?

NateMeyer commented:
I've played with a few different ones from the Intel OpenModelZoo repo.  I just use the omz_download and omz_convert to create an FP16 IR model with xml and bin files.  Any model that has the same output tensor shape as the ssdlite_mobilenet_v2 mode should just drop in.

There is also the model conversion CarlosML posted above, but that needs different post-processing than what I have.  Once I get my dev environment working again, I'll try using that since it is using the same model used by the edgetpu detector.

NateMeyer committed:
Print all available output tensors

NateMeyer commented:
Rebased on updates to detector-abstraction (#3656).

NickM-27 commented:
I think you'll need to rebase this on the latest dev 👍

NateMeyer commented:
Running some tests before I push the rebase. Ran into an issue with the password encoding change.  I previously had the password encoded in my config, putting special characters in is causing an issue. Will open an issue once I make down the problem.

NateMeyer committed:
Update documentation for config parameters

EduardoLaranjo commented:
hi @NateMeyer just wanted to say amazing work 👍

NateMeyer commented:
> One issue I was unable run openvino with my GPU even thought it's supported.
> 
> I did some investigation with `hello_query_device.py`, without any changes to the container, it only showed the CPU

Hey @EduardoLaranjo, what does your detector and model config look like?   What error do you get when you try to use the GPU detector?  How are you passing the GPU to docker? 

> After I install apt-get install intel-opencl-icd I could see both the GPU and CPU

Did you install this in the docker container or on your host?

EduardoLaranjo commented:
**what does your detector and model config look like?**
```
detectors:
  ov_detector:
    type: openvino
    device: GPU

ffmpeg:
  hwaccel_args: -c:v h264_qsv

model:
  path: /openvino-model/ssdlite_mobilenet_v2.xml
  width: 300
  height: 300
  input_tensor: nhwc
  input_pixel_format: bgr
  labelmap_path: /openvino-model/coco_91cl_bkgr.txt
```

**What error do you get when you try to use the GPU detector?** 
```
Traceback (most recent call last):
  File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/frigate/frigate/object_detection.py", line 117, in run_detector
    object_detector = LocalObjectDetector(
  File "/opt/frigate/frigate/object_detection.py", line 62, in __init__
    self.detect_api = OvDetector(
  File "/opt/frigate/frigate/detectors/openvino.py", line 16, in __init__
    self.interpreter = self.ov_core.compile_model(
  File "/usr/local/lib/python3.9/dist-packages/openvino/runtime/ie_api.py", line 387, in compile_model
    super().compile_model(model, device_name, {} if config is None else config),
RuntimeError: Failed to create plugin /usr/local/lib/python3.9/dist-packages/openvino/libs/libopenvino_intel_gpu_plugin.so for device GPU
Please, check your environment
[CLDNN ERROR]. clGetPlatformIDs error -1001
```

**How are you passing the GPU to docker?**
```
    devices:
      - /dev/dri/card0:/dev/dri/card0
      - /dev/dri/renderD128:/dev/dri/renderD128
```

Probably the most different thing in my docker file would be:
` entrypoint: bash -c "echo 'deb http://deb.debian.org/debian testing main non-free' >> /etc/apt/sources.list; apt -y update; apt -y -t testing install intel-media-va-driver-non-free; /init python3 -u -m frigate"`

Did you install this in the docker container or on your host?
Inside docker container

EduardoLaranjo commented:
hi @NateMeyer override the entrypoint with 

`    entrypoint: bash -c "echo 'deb http://deb.debian.org/debian testing main non-free' >> /etc/apt/sources.list; apt -y update; apt -y -t testing install intel-media-va-driver-non-free; apt -y install intel-opencl-icd; /init python3 -u -m frigate"`

Made it worked

NateMeyer commented:
You're installing those from the testing repo?  What version is getting installed?  This is what I see in the current image

```bash
$ docker run ghcr.io/natemeyer/frigate:0.12.0-openvino-8230b34 apt list --installed | grep intel-

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

intel-media-va-driver-non-free/now 21.1.1+ds1-1 amd64 [installed,local]
intel-opencl-icd/now 20.44.18297-1 amd64 [installed,local]

```

NickM-27 commented:
Looks good, we'll definitely want to update the docs recommended hardware to users know that these options exist, but I think it makes sense to wait given Nvidia Tensor devices will probably be added to that as well.

Only other thought is might want to look at https://github.com/blakeblackshear/frigate/pull/4395 and see if that can also work for the stages that are being added in this PR

NateMeyer commented on docker/Dockerfile:
```diff
@@ -42,6 +42,52 @@ RUN pip3 install -r requirements.txt
+
+ARG DEBIAN_FRONTEND="noninteractive"
+
+# Build libUSB without udev.  Needed for Openvino NCS2 support
```
No, I don't have a coral to test with.  That would be a good thing to check.

NickM-27 replied:
I only have PCIe corals, but I think I know someone who has a USB coral and could test. Do you have an image with the latest code from here?

NateMeyer replied:
Yea, this image should have the latest `ghcr.io/natemeyer/frigate:0.12.0-openvino-8230b34`

NickM-27 replied:
Thanks, I found a decent priced USB coral on amazon so worst case in a couple days I'll put it on my test Pi and see how ti goes.

NickM-27 replied:
@NateMeyer I'll have to do a bit more testing but unfortunately with this image the usb coral fails to be initialized

```
test_frigate    | F driver/usb/usb_driver.cc:857] transfer on tag 2 failed. Abort. Deadline exceeded: USB transfer error 2 [LibUsbDataOutCallback]
test_frigate    | Fatal Python error: Aborted
test_frigate    | 
test_frigate    | Thread 0x0000ffffa97771e0 (most recent call first):
test_frigate    |   File "/usr/lib/python3.9/threading.py", line 312 in wait
test_frigate    |   File "/usr/lib/python3.9/multiprocessing/queues.py", line 233 in _feed
test_frigate    |   File "/usr/lib/python3.9/threading.py", line 892 in run
test_frigate    |   File "/usr/lib/python3.9/threading.py", line 954 in _bootstrap_inner
test_frigate    |   File "/usr/lib/python3.9/threading.py", line 912 in _bootstrap
test_frigate    | 
test_frigate    | Thread 0x0000ffffb5b0f010 (most recent call first):
test_frigate    |   File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 833 in invoke
test_frigate    |   File "/opt/frigate/frigate/detectors/edgetpu_tfl.py", line 40 in detect_raw
test_frigate    |   File "/opt/frigate/frigate/object_detection.py", line 88 in detect_raw
test_frigate    |   File "/opt/frigate/frigate/object_detection.py", line 144 in run_detector
test_frigate    |   File "/usr/lib/python3.9/multiprocessing/process.py", line 108 in run
test_frigate    |   File "/usr/lib/python3.9/multiprocessing/process.py", line 315 in _bootstrap
test_frigate    |   File "/usr/lib/python3.9/multiprocessing/popen_fork.py", line 71 in _launch
test_frigate    |   File "/usr/lib/python3.9/multiprocessing/popen_fork.py", line 19 in __init__
test_frigate    |   File "/usr/lib/python3.9/multiprocessing/context.py", line 277 in _Popen
test_frigate    |   File "/usr/lib/python3.9/multiprocessing/context.py", line 224 in _Popen
test_frigate    |   File "/usr/lib/python3.9/multiprocessing/process.py", line 121 in start
test_frigate    |   File "/opt/frigate/frigate/object_detection.py", line 205 in start_or_restart
test_frigate    |   File "/opt/frigate/frigate/object_detection.py", line 174 in __init__
test_frigate    |   File "/opt/frigate/frigate/app.py", line 204 in start_detectors
test_frigate    |   File "/opt/frigate/frigate/app.py", line 358 in start
test_frigate    |   File "/opt/frigate/frigate/__main__.py", line 16 in <module>
test_frigate    |   File "/usr/lib/python3.9/runpy.py", line 87 in _run_code
test_frigate    |   File "/usr/lib/python3.9/runpy.py", line 197 in _run_module_as_main
```

NickM-27 replied:
@NateMeyer After restarting my Pi it finds the USB coral so seems it might have just been overloaded for some reason.

NateMeyer replied:
Interesting.  Glad to hear it worked.

NateMeyer commented:
@EduardoLaranjo I believe the updated driver support will be included in #4368

NickM-27 commented:
@NateMeyer this will need to be rebased on dev since there have been a bunch of docker changes. Should make for a cleaner build though

NickM-27 commented:
> Yes, I've been watching for it to settle out. Are there any more issue or changes for the build/docker/dev env needed or should it be ready now?

I think there may be a few stragglers but I think (hope) the structure should be finished and no more organizational changes

NateMeyer commented:
Rebased. Built a new test image at `ghcr.io/natemeyer/frigate:0.12.0-openvino-fdcc476`

nickp27 commented:
@NateMeyer love your work, this has been amazing on my intel thin client for lowering power and CPU usage. One thing that has plagued me is that ssdlite_mobilenet_v2 seems to be causing a ridiculous number of false positives picking up a a firepit and some pot plants as human, but at such a high percentage that I can't eliminate it with thresholds (see below):

![Capture](https://user-images.githubusercontent.com/2366809/205434453-b44edd89-784d-486e-b735-6c5589871bde.PNG)

Any guides or instructions on what I need to change in the Dockerfile to test another model (I understand from the above that anything with the same post-processing can just be swapped out) - I assume by simply changing the omz-download and omz-converter lines.

blakeblackshear commented:
Is there any outstanding work for this PR, or is it ready to merge? Have you tested a multiarch build?

NateMeyer commented:
> Is there any outstanding work for this PR, or is it ready to merge? Have you tested a multiarch build?

I think this is ready.  The only outstanding work may be tweaking the included model, as @nickp27 points out above.  I have been building the multiarch image, but I've only run it with qemu personally (amd64, arm64, and arm32).  There was some work in  #3797 to run it with the NCS2 stick on amd64 and arm64.  The image seemed to work ok on arm64, but I think we were running into some issues with the NCS2 on kubernetes.

> Any guides or instructions on what I need to change in the Dockerfile to test another model (I understand from the above that anything with the same post-processing can just be swapped out) - I assume by simply changing the omz-download and omz-converter lines.

@nickp27, Clearly you need to move your plants to not be so lifelike :laughing: .  I'll dig up my notes on which models dropped in.  If this gets merged I think you can open a regular issue and may get other suggestions about masking it out, I'm not as familiar myself with creating zones.  Are you thinking of rebuilding the image or you can run just the ov-converter stage and mount the converted model in the prebuilt image.

NateMeyer commented:
@nickp27 here are a list of some models I think will drop in from the [Intel OMZ.](https://github.com/openvinotoolkit/open_model_zoo/blob/master/models/public/index.md)

- ssdlite_mobilenet_v2 (default)
- mobilenet-ssd
- efficientdet-d0-tf
- ssd300

These models may have different sizes and labelmaps that also need to have the correct adjustments in the model config.  Also, I didn't have much luck with the ssd300 model, I didn't debug why I wasn't getting detection boxes even though the debug log seemed to have detections.  Others may be the same.

I wanted to try a nanodet or yolo model, but this work was left for another time.

nickp27 commented:
@NateMeyer Thanks for that! I tried converting yolov4-tiny to match the development work being done on tensorrt but haven't figured out how to parse the multiple outputs just yet. Will give a combination of new models and some filters a go!

thekillerpt commented:
not sure if this is the right place, just to provide some feedback,
HA running on a laptop 2630qm , I7, 2nd generation sandy bridge cpu, 5 reolink cameras detecting at 640x352 and capturing at 2k
tried to enable open vino just by using the config example on beta docs just to give it a go 

cpu util dropped from 60% (at most, had only 4 cores allowed in detectors) to 25% (at most)

inference speed dropped from 100ms to 75ms 

thanks a lot for this ! :-) no need for coral nor machine upgrade for now
